src/execution_plan/execution_plan_build/execution_plan_construct.c

```diff
@@ -233,6 +233,7 @@ static void _buildForeachOp
 	ExecutionPlan *embedded_plan = ExecutionPlan_NewEmptyExecutionPlan();
 	embedded_plan->ast_segment = body_ast;
 	embedded_plan->record_map = raxClone(plan->record_map);
+	const char **arguments = (const char **)raxKeys(embedded_plan->record_map);
```

Comment:
Dangerous, for the reason where `record_map` might include variables which are bounded after the `FOREACH` clause. Moreover (just making sure) the `arguments` list should include the loop variable e.g. `FOREACH p IN posts` should `p` be in `arguments`?

Reply:
in this location record_map contains only the variables befor the foreach is constructed
p is in the unwind modified

Reply:
I think this is a bit dangerous to always assume, for safety reasons can we switch to do bound entities collection via execution-plan traversal?